app/src/main/java/com/github/jorgecastillo/kotlinandroid/data/datasource/remote/MarvelNetworkDataSource.kt

```diff
@@ -24,35 +25,31 @@ import java.net.HttpURLConnection
  * word in the title. Yep, I wanted to retrieve Avengers but the Marvel API is a bit weird
  * sometimes.
  */
+fun exceptionAsCharacterError(e: Throwable): CharacterError =
+        when (e) {
+            is MarvelAuthApiException -> AuthenticationError
+            is MarvelApiException ->
+                if (e.httpCode == HttpURLConnection.HTTP_NOT_FOUND) NotFoundError
+                else UnknownServerError(Option.Some(e))
+            else -> UnknownServerError((Option.Some(e)))
+        }
 
-fun fetchAllHeroes() = Reader.ask<GetHeroesContext>().map { ctx ->
-  Future {
-    try {
-      val query = CharactersQuery.Builder.create().withOffset(0).withLimit(50).build()
-      Right<List<CharacterDto>>(ctx.apiClient.getAll(query).response.characters)
-    } catch (e: MarvelAuthApiException) {
-      Left(AuthenticationError())
-    } catch (e: MarvelApiException) {
-      if (e.httpCode == HttpURLConnection.HTTP_NOT_FOUND) {
-        Left(NotFoundError())
-      } else {
-        Left(UnknownServerError())
-      }
-    }
-  }
-}
 
-fun fetchHeroesFromAvengerComics() = fetchAllHeroes().map { future ->
-  future.map {
-    when (it) {
-      is Right -> it.map {
-        it.filter {
-          it.comics.items.map { it.name }.filter {
-            it.contains("Avenger", true)
-          }.count() > 0
+inline fun <reified F> fetchAllHeroes(C : Control<F> = control()): HK<F, List<CharacterDto>> =
+        C.binding {
+            val query = CharactersQuery.Builder.create().withOffset(0).withLimit(50).build()
+            val ctx = !C.ask()
```

Comment:
does the `!` stand for `.bind()` ? I would rather using an easier to read / understand operator / method here if we have it, since the `!` does not really mean anything to the reader of this code. This is a didactic project so I really would like people to not have a hard time understanding the code. Let's also think that people reading this are not gonna be familiarized with FP almost at all.